satellite/metabase/rangedloop/service.go

```diff
@@ -74,81 +86,116 @@ func (service *Service) Run(ctx context.Context) (err error) {
 }
 
 // RunOnce goes through one time and sends information to observers.
-func (service *Service) RunOnce(ctx context.Context) (err error) {
+func (service *Service) RunOnce(ctx context.Context) (observerDurations []ObserverDuration, err error) {
 	defer mon.Task()(&ctx)(&err)
 
-	startTime := time.Now()
-	observerStates := []observerState{}
-
-	for _, obs := range service.observers {
-		err := obs.Start(ctx, startTime)
-		if err != nil {
-			return err
-		}
-
-		observerStates = append(observerStates, observerState{
-			observer: obs,
-		})
+	observerStates, err := startObservers(ctx, service.observers)
+	if err != nil {
+		return nil, err
 	}
 
 	rangeProviders, err := service.provider.CreateRanges(service.config.Parallelism, service.config.BatchSize)
 	if err != nil {
-		return err
+		return nil, err
 	}
 
 	group := errs2.Group{}
-	for rangeIndex, rangeProvider := range rangeProviders {
-		rangeObservers := []Partial{}
+	for _, rangeProvider := range rangeProviders {
+		rangeObservers := []*rangeObserverState{}
 		for i, observerState := range observerStates {
 			rangeObserver, err := observerState.observer.Fork(ctx)
 			if err != nil {
-				return err
+				return nil, err
 			}
-			rangeObservers = append(rangeObservers, rangeObserver)
-			observerStates[i].rangeObservers = append(observerStates[i].rangeObservers, rangeObserver)
-		}
-
-		// Create closure to capture loop variables.
-		createClosure := func(ctx context.Context, rangeIndex int, rangeProvider SegmentProvider, rangeObservers []Partial) func() error {
-			return func() (err error) {
-				defer mon.Task()(&ctx)(&err)
-
-				return rangeProvider.Iterate(ctx, func(segments []segmentloop.Segment) error {
-					for _, rangeObserver := range rangeObservers {
-						err := rangeObserver.Process(ctx, segments)
-						if err != nil {
-							return err
-						}
-					}
-					return nil
-				})
+			rangeState := &rangeObserverState{
+				rangeObserver: rangeObserver,
+				duration:      time.Duration(0),
```

Comment:
nitpick: don't need to initialize this field since the zero value is identical to the value being explicitly initialized.

Reply:
Done